zitadel/v2/instance_member/funcs.go

```diff
@@ -0,0 +1,143 @@
+		UserId: d.Get(userIDVar).(string),
+	})
+	if err != nil {
+		return diag.Errorf("failed to delete instance member: %v", err)
+	}
+	return nil
+}
+
+func update(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	tflog.Info(ctx, "started update")
+
+	clientinfo, ok := m.(*helper.ClientInfo)
+	if !ok {
+		return diag.Errorf("failed to get client")
+	}
+
+	client, err := helper.GetAdminClient(clientinfo)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	_, err = client.UpdateIAMMember(ctx, &admin.UpdateIAMMemberRequest{
+		UserId: d.Get(userIDVar).(string),
+		Roles:  helper.GetOkSetToStringSlice(d, rolesVar),
+	})
+	if err != nil {
+		return diag.Errorf("failed to update instance member: %v", err)
+	}
+	return nil
+}
+
+func create(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	tflog.Info(ctx, "started create")
+
+	clientinfo, ok := m.(*helper.ClientInfo)
+	if !ok {
+		return diag.Errorf("failed to get client")
+	}
+
+	client, err := helper.GetAdminClient(clientinfo)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	userID := d.Get(userIDVar).(string)
+	resp, err := client.AddIAMMember(ctx, &admin.AddIAMMemberRequest{
+		UserId: userID,
+		Roles:  helper.GetOkSetToStringSlice(d, rolesVar),
+	})
+	if err != nil {
+		return diag.Errorf("failed to create instance member: %v", err)
+	}
+	d.SetId(getInstanceMemberID(resp.GetDetails().GetResourceOwner(), userID))
+	return nil
+}
+
+func read(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	tflog.Info(ctx, "started read")
+
+	clientinfo, ok := m.(*helper.ClientInfo)
+	if !ok {
+		return diag.Errorf("failed to get client")
+	}
+
+	client, err := helper.GetAdminClient(clientinfo)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	userID := d.Get(userIDVar).(string)
+	resp, err := client.ListIAMMembers(ctx, &admin.ListIAMMembersRequest{
+		Queries: []*member.SearchQuery{{
+			Query: &member.SearchQuery_UserIdQuery{
+				UserIdQuery: &member.UserIDQuery{
+					UserId: userID,
+				},
+			},
+		}},
+	})
+	if err != nil && helper.IgnoreIfNotFoundError(err) == nil {
+		d.SetId("")
+		return nil
+	}
+	if err != nil {
+		return diag.Errorf("failed to list instance members")
+	}
+
+	if len(resp.Result) == 1 {
+		member := resp.Result[0]
+		set := map[string]interface{}{
+			userIDVar: member.GetUserId(),
+			rolesVar:  member.GetRoles(),
+		}
+		for k, v := range set {
+			if err := d.Set(k, v); err != nil {
+				return diag.Errorf("failed to set %s of instance member: %v", k, err)
+			}
+		}
+		d.SetId(getInstanceMemberID(member.GetDetails().GetResourceOwner(), userID))
+		return nil
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func getInstanceMemberID(instance string, userID string) string {
+	return instance + "_" + userID
+}
+
+func splitInstanceMemberID(memberID string) (string, string) {
+	parts := strings.Split(memberID, "_")
+	return parts[0], parts[1]
+}
```

Comment:
Do we really need to namespace by the instance, aren't the IDs unique anyways?
If they aren't unique, we should require the instance ID as input, IMO.

Reply:
technically you can mutliple zitadel providers in the same folder which share the state, so if the providers are for different instances it is necessary to have a unique ID